language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module defines the (acyclic) borrow graph for the type and memory safety analysis.
+use crate::nonce::Nonce;
+use mirai_annotations::checked_verify;
+use std::collections::{BTreeMap, BTreeSet};
+
+/// There are two types of edges in the borrow graph.
+/// A strong edge carries more information than a weak edge.
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum EdgeType {
+    Strong,
+    Weak,
+}
+
+/// The label on an edge
+pub type Label<T> = Vec<T>;
+
+/// A labeled edge in borrow graph
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Edge<T> {
+    edge_type: EdgeType,
+    label: Label<T>,
+    to: Nonce,
+}
+
+/// A borrow graph is represented as a map from a source nonce to the set of all edges
+/// coming out of it.
+#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct BorrowGraph<T>(BTreeMap<Nonce, BTreeSet<Edge<T>>>);
+
+impl<T> BorrowGraph<T>
+where
+    T: Ord,
+    T: Clone,
+{
+    /// creates a new empty borrow graph
+    pub fn new() -> Self {
+        BorrowGraph(BTreeMap::new())
+    }
+
+    /// adds a fresh nonce
+    pub fn add_nonce(&mut self, nonce: Nonce) {
+        checked_verify!(!self.0.contains_key(&nonce));
+        self.0.insert(nonce, BTreeSet::new());
+    }
+
+    /// adds a weak edge
+    /// `to` must be a fresh nonce
+    pub fn add_weak_edge(&mut self, from: Nonce, label: Label<T>, to: Nonce) {
+        checked_verify!(self.0.contains_key(&from));
+        checked_verify!(self.0.contains_key(&to));
```

Comment:
You say `to` must be fresh nonce, but here you require it to be already contained in the graph. Something does not sound right here. At the very least you are being inconsistent with line 45 above.

Reply:
See comment above.  All I wanted was that there shouldn't be any outgoing edges out of `to`.